examples/monitoring_driven_software_defined_assets/docs/decisions/0003-start_by_monitorng_asset_quality.md

```diff
@@ -0,0 +1,63 @@
+# 8. Start by monitoring asset quality
+
```

Comment:
@sryza After going around in circles a bit I've refocussed to focus this example on the technique of building the *monitoring* of the Asset before building the asset as a way to focus development and ensure that a data asset is delivered along with a monitoring that validates its quality.

I'd love to hear your high level thoughts on using a separate "monitoring" Dagster job to validate the quality of Assets.

Reply:
Your understanding is spot on!

> how do you want the production jobs to work? I.e. do you really want the data quality checks to run in a job that's independent from the job that refreshes the asset? Would you not want to re-run the data quality checks each time the asset is re-generated?

Excellent questions... and currently I'm unsure.  Let's discuss.

Philosophically I think that a monitoring system should be in a different failure domain to the system being monitored - ie; I don't want the monitoring to fail for the same reason that the thing being monitored fails.

For example; should the "scheduling system" fail to trigger an update of the data asset I want the monitoring to start failing because the asset has stale data.

I think this requires the data quality checks to run in an independent job (possibly running in an independent Dagster deployment?)

I don't see any harm in *also* running the data quality checks each time the asset is re-generated; since this would provide the quickest possible detection of a problem.  However, I don't think this is the only place they can run since it seems that there would be many failure cases that would impact both the data generating job and the data monitoring job.

WDYT?

Reply:
That makes sense. That also fits with doing data quality checks and monitoring on foreign assets, for which it wouldn't make sense to run checks in the same run where they're created (because they aren't created in any run).

So that then makes me wonder what the ideal API looks like for supporting this.

AssetObservations should land soon, and then you'll be able to write a job that logs arbitrary metadata about any asset. A drawback there is that the data quality checks end up disconnected from the asset definition, so you can't view what's expected of an asset without running the monitoring job.

To unite the two, it might make sense to provide an API constructs a monitoring op from an AssetDefinition.  E.g.

```
from dagster import asset, DataFrameDagsterType, job, create_monitor_asset_op

@asset(dagster_type=DataFrameDagsterType(column_constraints...))
def asset1():
    ...

monitor_asset1 = create_monitor_asset_op(asset1)

@job
def my_job():
    monitor_asset1()
```

Thoughts?